C3DC-1936 Update components with kick out feature

huuaho opened the pull request:
## Description

Ported over the new kick out feature from CCDI Hub 

Fixes [C3DC-1936](https://tracker.nci.nih.gov/browse/C3DC-1936)

## Type of change
- Added new components as seen in CCDI-Hub
- Added extended style customization for components from the frontend (on C3DC each section customizes its color to its sub components such as the slider and search facet buttons)
- The sliders have been overhauled to have different style customizations + unknown age support + toggle unit
- Query bar has added support to show the different states of unknownAges
- When the kickout expands, it encroaches into the main explore page. This shrunk the space of the tabs but the tabs did not respond. Modified tabs to respond to width of parent div as opposed to width of the screen.


CHANGES CAPTURED FROM OTHER TICKETS:
[C3DC-1910](https://tracker.nci.nih.gov/browse/C3DC-1910) 
- Moved the divider to be below the related facet section
- As a part of some other style changes we enabled the bar charts to accept custom colors to update the widgets with new colors

[C3DC-1980](https://tracker.nci.nih.gov/browse/C3DC-1980)
- Added support for unit toggle (day/year) and unknown age data

Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)


## How Has This Been Tested?
Locally

huuaho committed:
added new actions

huuaho committed:
added new components

huuaho committed:
update package vers

huuaho committed:
added ability to customize the search facet button

huuaho committed:
customized search facet positioning

huuaho committed:
queryparams slider

huuaho committed:
added new sliders

huuaho committed:
adding custom configs for slider button styling

huuaho committed:
removed extra bounds in sliders

huuaho committed:
Merge branch '1.0.1-c3dc' into C3DC-1936

huuaho committed:
fixed comment typo

huuaho commented on packages/facet-filter/src/NewFacetFilterController.js:
```diff
@@ -0,0 +1,184 @@
+              return addField;
+            });
+            /**
+             * Check if there are orphen filter values and add them to the facet values
```
ok

huuaho committed:
query bar support for unknown ages

huuaho committed:
Merge branch '1.0.1-c3dc' into C3DC-1936

huuaho committed:
allowed for bar charts to use custom colors

huuaho committed:
updated tab to be responsive to kickout panel

huuaho committed:
reduced null risks

huuaho committed:
removed constructed query structures

amattu2 commented on packages/facet-filter/src/components/facet/NewFacetView.js:
```diff
@@ -0,0 +1,266 @@
+              <ToggleButtonGroup
+                value={timeUnit}
+                exclusive
+                disabled={isOnlyUnknownAges}
```
Does `disabled` exist on the `ToggleButtonGroup`? 

https://v4.mui.com/api/toggle-button-group/

```suggestion

```

huuaho replied:
Also pulled from CCDI Hub
https://github.com/CBIIT/bento-frontend/blob/24ae877d9dd774acf06419256b13e9481325bd1d/packages/facet-filter/src/components/facet/NewFacetView.js#L195

No, after checking we have to pass disabled via the children components (ToggleButton) which we also do. this 'disabled' field is redundant - or rather doesn't actually function.

huuaho replied:
removed

amattu2 commented on packages/facet-filter/src/components/facet/NewSearchFacetView.js:
```diff
@@ -0,0 +1,299 @@
+                      right: 7,
+                      top: 7,
+                      color: theme.palette.grey[500],
+                    })}
```
Does `sx` exist in Material UI v4? I thought thats v5 or above. The only reason I'm pointing this out is because my IDE is complaining about invalid props.

https://v4.mui.com/api/icon-button/#iconbutton-api

huuaho replied:
Also pulled from CCDI Hub https://github.com/CBIIT/bento-frontend/blob/24ae877d9dd774acf06419256b13e9481325bd1d/packages/facet-filter/src/components/facet/NewSearchFacetView.js#L217

Also found in two other locations in our current codebase - these were also pulled from CCDI Hub when adding in the 'Search Facets feature'

https://github.com/CBIIT/bento-frontend/blob/98699b3b5ff7518f77b476e7c0ffcc9619787b20/packages/facet-filter/src/components/facet/FacetModal.js#L51
https://github.com/CBIIT/bento-frontend/blob/98699b3b5ff7518f77b476e7c0ffcc9619787b20/packages/facet-filter/src/components/facet/SearchFacetView.js#L161

huuaho replied:
No sx does not exist in mui v4

And i just checked that CCDI Hub and C3DC are using the same versions.

huuaho replied:
removed from all locations

amattu2 commented on packages/facet-filter/src/components/inputs/slider/SliderView.js:
```diff
@@ -3,8 +3,24 @@
+  useNavigate,
+} from 'react-router-dom';
+*/
+import { generateQueryStr } from '@bento-core/util';
```
Unused import for now


```suggestion
// import { generateQueryStr } from '@bento-core/util';
```

huuaho replied:
good catch!

amattu2 commented on packages/facet-filter/src/components/inputs/slider/SliderView.js:
```diff
@@ -320,6 +521,72 @@ const styles = () => ({
+      color: '#000000',
+    }),
+  radio: (props) => {
+    const primaryColor = (props.facet && props.facet.style && props.facet.style.colorPrimary)
```
Unused `primaryColor`. Is this intentional?

huuaho replied:
https://github.com/CBIIT/bento-frontend/blob/24ae877d9dd774acf06419256b13e9481325bd1d/packages/facet-filter/src/components/inputs/slider/SliderView.js#L553
Pulled over from CCDI, it is also unused there. We can remove.

amattu2 commented on packages/facet-filter/src/NewFacetFilterController.js:
```diff
@@ -0,0 +1,184 @@
+/* eslint-disable no-restricted-syntax */
+/* eslint-disable no-param-reassign */
+/* eslint-disable space-infix-ops */
+/* eslint-disable prefer-template */
```
Are any of these rules actually enforced? I assume it was just copy-pasted over. Let's remove them if they aren't needed.

```suggestion

```

huuaho replied:
yes they are also pulled from CCDI Hub. Anything with the 'New' components pretty much was pulled from them directly

huuaho replied:
Also directly pulled https://github.com/CBIIT/bento-frontend/blob/24ae877d9dd774acf06419256b13e9481325bd1d/packages/facet-filter/src/NewFacetFilterController.js#L1 

went through one by one and removed all the eslint issues.

It actually improved efficiency a lot and the components load way quicker, go figure

amattu2 commented on packages/facet-filter/package.json:
```diff
@@ -16,15 +16,18 @@
     "access": "public"
   },
   "dependencies": {
+    "@bento-core/util": "^1.0.0-c3dc.0",
```
Can you confirm if we still need this dependency? I assume it's from the query string portion?

huuaho replied:
Yes it is, last year I had modified util.js to house the generateQueryStr func to have parity with CCDI Hub. (This is the only difference) Since we are using it in facet-filter in the future and most of the 'New' components was pulled from their code I just kept it since we're considering to use it in the future.

huuaho replied:
removed (for now)

amattu2 reviewed (changes requested):
Generally looks good, I pointed out a few things we need to look into. LMK if you have any questions

huuaho committed:
removed unnecessary definitions

huuaho committed:
eliminate parameter mutations in NewFacetFilterController

  Replace forEach loops and direct property assignments with immutable
  patterns using map/reduce and object spread operators to comply with
  no-param-reassign eslint rule

huuaho committed:
use immutable patterns and remove eslint violations

  Replace parameter mutations with map/reduce and spread operators to
  comply with no-param-reassign rule. Optimize updateFacetState to use
  direct property access instead of iterating through all filter entries.

huuaho committed:
removed unused. package
